dot/core/service_test.go

```diff
@@ -0,0 +1,616 @@
+	require.NoError(t, err)
+
+	// hash for known test code substitution
+	blockHash := common.MustHexToHash("0x86aa36a140dfc449c30dbce16ce0fea33d5c3786766baa764e33f336841b9e29")
+	testCodeSubstitute := map[common.Hash]string{
+		blockHash: common.BytesToHex(testRuntime),
+	}
+
+	runtimeMock := new(mocksruntime.Instance)
+	runtimeMock.On("Keystore").Return(&keystore.GlobalKeystore{})
+	runtimeMock.On("NodeStorage").Return(runtime.NodeStorage{})
+	runtimeMock.On("NetworkService").Return(new(runtime.TestRuntimeNetwork))
+	runtimeMock.On("Validator").Return(true)
+
+	ctrl := gomock.NewController(t)
+	mockBlockStateGetRtErr := NewMockBlockState(ctrl)
+	mockBlockStateGetRtErr.EXPECT().GetRuntime(gomock.Any()).Return(nil, errTestDummyError)
+
+	mockBlockStateGetRtOk1 := NewMockBlockState(ctrl)
+	mockBlockStateGetRtOk1.EXPECT().GetRuntime(gomock.Any()).Return(runtimeMock, nil)
+
+	mockBlockStateGetRtOk2 := NewMockBlockState(ctrl)
+	mockBlockStateGetRtOk2.EXPECT().GetRuntime(gomock.Any()).Return(runtimeMock, nil)
+	mockBlockStateGetRtOk2.EXPECT().StoreRuntime(blockHash, gomock.Any())
+
+	mockCodeSubState1 := NewMockCodeSubstitutedState(ctrl)
+	mockCodeSubState1.EXPECT().StoreCodeSubstitutedBlockHash(blockHash).Return(errTestDummyError)
+
+	mockCodeSubState2 := NewMockCodeSubstitutedState(ctrl)
+	mockCodeSubState2.EXPECT().StoreCodeSubstitutedBlockHash(blockHash).Return(nil)
+
+	type args struct {
+		hash  common.Hash
+		state *rtstorage.TrieState
+	}
+	tests := []struct {
+		name      string
+		service   *Service
+		args      args
+		expErr    error
+		expErrMsg string
+	}{
+		{
+			name:    "nil value",
+			service: &Service{codeSubstitute: map[common.Hash]string{}},
+			args: args{
+				hash: common.Hash{},
+			},
+		},
+		{
+			name: "getRuntime error",
+			service: &Service{
+				codeSubstitute: testCodeSubstitute,
+				blockState:     mockBlockStateGetRtErr,
+			},
+			args: args{
+				hash: blockHash,
+			},
+			expErr:    errTestDummyError,
+			expErrMsg: errTestDummyError.Error(),
+		},
+		{
+			name: "code substitute error",
+			service: &Service{
+				codeSubstitute:       testCodeSubstitute,
+				blockState:           mockBlockStateGetRtOk1,
+				codeSubstitutedState: mockCodeSubState1,
+			},
+			args: args{
+				hash: blockHash,
+			},
+			expErr:    errTestDummyError,
+			expErrMsg: errTestDummyError.Error(),
+		},
+		{
+			name: "happyPath",
+			service: &Service{
+				codeSubstitute:       testCodeSubstitute,
+				blockState:           mockBlockStateGetRtOk2,
+				codeSubstitutedState: mockCodeSubState2,
+			},
+			args: args{
+				hash: blockHash,
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := tt.service
+			err = s.handleCodeSubstitution(tt.args.hash, tt.args.state)
+			assert.ErrorIs(t, err, tt.expErr)
+			if tt.expErr != nil {
+				assert.EqualError(t, err, tt.expErrMsg)
+			}
+		})
+	}
+}
+
+func TestService_handleBlock(t *testing.T) {
```

Comment:
Improve readability with 2 ways:

### Mocks configuration changing a lot from a test case to another

⚠️ this is usually a code smell showing the production function being tested is too long/complex.

BUT it does happen. This is the case here.

We don't want to have huge mocks definition away from a particular test case also away from the actual subtest testing code, this hurts readability. We can't either define mocks from test cases in subtests since this would need a bunch of structs and if blocks in the subtest, making unreadable as well.

The solution is to have just subtests **without** test cases slices or for loop, so everything is locally scoped despite a bit more code duplication.

Example:

```go
t.Run("store trie error", func(t *testing.T) {
	mockStorageStateErr := NewMockStorageState(ctrl)
	mockStorageStateErr.EXPECT().StoreTrie(trieState, &block.Header).Return(errTestDummyError)
	// ...
	s := &Service{storageState: mockStorageStateErr}
	err := s.handleBlock(block, state)
	assert.ErrorIs(t, err, expErr)
	if expErr != nil {
		assert.EqualError(t, err, expErrMsg)
	}
})
```

### Mocks configuration similar from a test case to another

Then configure the mocks in the subtest, and have its expected args / return values in the test cases slice.

Reply:
I like this. I think this would be a good "unit test standard" to help out tests be more consistent across the codebase

Reply:
Added in https://github.com/ChainSafe/gossamer/pull/2311